src/sql/src/session/vars.rs

```diff
@@ -568,16 +582,22 @@ impl SessionVars {
         action: EndTransactionAction,
     ) -> BTreeMap<&'static str, String> {
         let mut changed = BTreeMap::new();
-        for var in self.vars.values_mut() {
+        let mut updates = Vec::new();
+        for (name, var) in self.vars.iter() {
+            if !var.is_mutating() {
+                continue;
+            }
             let before = var.value();
-            var.end_transaction(action);
-            let after = var.value();
+            let next = var.end_transaction(action).expect("must mutate");
+            let after = next.value();
+            updates.push((*name, next));
```

Comment:
Could this be refactored to:
```
let before = var.value();
if let Some(next) = var.end_transaction(action) {
  let after = next.value();
  updates.push((*name, next));
  
  if before != after {
    changed.insert(var.name(), after);
  }
}
```
which would move the `var.is_mutating()` check into a single place

Reply:
Yes, but that also requires calling `self.value_dyn().format()` and allocating a String for *every* session var on every transaction end (which also happens implicitly for things like `select 1`). I think the extra if block is well worth the slight increase in code lines to avoid so much allocation.

Reply:
Ah ha! Yeah good call